File: src/binary_search/zig/binary_search.zig
pub fn binary_search(comptime T: type, arr: []T, key: T) i32 {
    var left: u32 = 0;
    var right: u32 = @intCast(u32, arr.len);
    if (arr.len == 1) {
        if (arr[0] == key) return @intCast(i32, 0);
        return -1;
    }
    while (left < right) {
        var mid = (left + right) / 2;
        if (arr[mid] == key) return @intCast(i32, mid);
        if (arr[mid] < key) left = mid;
        if (arr[mid] > key) right = mid;
    }
    return -1;
}

pub fn main() !void {}

const assert = @import("std").debug.assert;

test "Empty List" {
    var arr = [_]u32{};
    assert(binary_search(u32, &arr, 1) == -1);
}

test "Element present in List" {
    var arr = [_]u32{ 1, 3, 5, 7 };
    assert(binary_search(u32, &arr, 3) == 1);
}

test "Element present at the end of the list" {
    var arr = [_]u32{ 1, 3, 5, 7, 9, 10 };
    assert(binary_search(u32, &arr, 10) == 5);
}

test "Element present in list with one element" {
    var arr = [_]u32{1};
    assert(binary_search(u32, &arr, 1) == 0);
}

test "Element absent in list with one element" {
    var arr = [_]u32{1};
    assert(binary_search(u32, &arr, 3) == -1);
}
test "List with negative numbers and negative key" {
    var arr = [_]i32{ -5, -4, -2, 0, 3, 5, 12 };
    assert(binary_search(i32, &arr, -2) == 2);
}
test "List with negative numbers and positive key" {
    var arr = [_]i32{ -5, -4, -2, 0, 3, 5, 12 };
    assert(binary_search(i32, &arr, 3) == 4);
}

File: src/binary_search/java/binary_search.java
import java.util.Scanner;
import java.io.File;
import java.io.FileNotFoundException;

class BinarySearch {
    public static <T extends Comparable<T>> int search(T[] arr, T key) {
        int left = 0;
        int right = arr.length;
        if (arr.length == 1) return arr[0].compareTo(key) == 0 ? 0 : -1;
        while (left < right) {
            int mid = (left + right) / 2;
            if (arr[mid].compareTo(key) == 0) return mid;
            if (arr[mid].compareTo(key) < 0) left = mid;
            if (arr[mid].compareTo(key) > 0) right = mid;
        }
        return -1;
    }

    private static void assertAndPrint(String testName, int testNumber, boolean expression) {
        System.out.printf(
            "Test %d: %s -> %s%n", testNumber, testName, expression ? "Passed" : "Failed"
        );
    }

    public static void main(String[] args) throws FileNotFoundException {
        assertAndPrint("Empty List", 1, 
        BinarySearch.search(new Integer[]{}, 0) == -1
    );
            assertAndPrint("Element present in List", 2,
        BinarySearch.search(new Integer[]{1, 3, 5, 7}, 3) == 1
    );
            assertAndPrint("Element present in list with one Element", 3,
        BinarySearch.search(new Integer[]{1}, 1) == 0
    );
            assertAndPrint("Element absent in list with one element", 4,
        BinarySearch.search(new Integer[]{1}, 3) == -1
    );
            assertAndPrint("Element present at the end of the list", 5,
        BinarySearch.search(new Integer[]{1, 3, 5, 7, 9, 10}, 10) == 5
    );
            assertAndPrint("List with negative numbers and negative key", 6,
        BinarySearch.search(new Integer[]{-5, -4, -2, 0, 3, 5, 12}, -2) == 2
    );
            assertAndPrint("List with negative numbers and positive key", 7,
        BinarySearch.search(new Integer[]{-5, -4, -2, 0, 3, 5, 12}, 3) == 4
    );
    }
}